File: src/core/llm/config.test.ts
/**
 * Tests for LLM config loading and provider factory
 *
 * These tests verify:
 * - Config file loading from ~/.config/memmem/config.json
 * - Provider creation from config
 * - Error handling for invalid configs
 */

import { describe, it, expect, beforeEach, afterEach, mock } from 'bun:test';
import { loadConfig, createProvider, __setConfigFileDepsForTests, DEFAULT_EMBEDDING_MAX_CONCURRENCY, type LLMConfig } from './config.js';

// File-dependency stubs (per-test controlled)
let mockExistsSyncReturnValue = false;
let mockReadFileSyncReturnValue = '{}';

const mockExistsSync = mock(() => mockExistsSyncReturnValue);
const mockReadFileSync = mock(() => mockReadFileSyncReturnValue);

describe('loadConfig', () => {
  beforeEach(() => {
    mockExistsSyncReturnValue = false;
    mockReadFileSyncReturnValue = '{}';
    mockExistsSync.mockClear();
    mockReadFileSync.mockClear();
    __setConfigFileDepsForTests({
      existsSync: mockExistsSync as unknown as typeof import('fs').existsSync,
      readFileSync: mockReadFileSync as unknown as typeof import('fs').readFileSync,
    });
  });

  afterEach(() => {
    __setConfigFileDepsForTests(null);
  });

  describe('when config file does not exist', () => {
    it('should return null', () => {
      const config = loadConfig();
      expect(config).toBeNull();
    });
  });

  describe('when config file exists and is valid', () => {
    it('should return LLMConfig with gemini provider', () => {
      const validConfig = {
        provider: 'gemini' as const,
        apiKey: 'test-api-key',
        model: 'gemini-2.0-flash',
      };
      mockExistsSyncReturnValue = true;
      mockReadFileSyncReturnValue = JSON.stringify(validConfig);

      const config = loadConfig();
      expect(config).toMatchObject(validConfig);
    });

    it('should return LLMConfig with zai provider', () => {
      const validConfig = {
        provider: 'zai' as const,
        apiKey: 'test-api-key',
        model: 'glm-4.7',
      };
      mockExistsSyncReturnValue = true;
      mockReadFileSyncReturnValue = JSON.stringify(validConfig);

      const config = loadConfig();
      expect(config).toMatchObject(validConfig);
    });

    it('should return LLMConfig with provider and apiKey (model defaults)', () => {
      const validConfig = {
        provider: 'gemini' as const,
        apiKey: 'test-api-key',
      };
      mockExistsSyncReturnValue = true;
      mockReadFileSyncReturnValue = JSON.stringify(validConfig);

      const config = loadConfig();
      expect(config).toMatchObject(validConfig);
    });

    it('should return LLMConfig with ratelimit settings', () => {
      const validConfig = {
        provider: 'gemini' as const,
        apiKey: 'test-api-key',
        ratelimit: {
          embedding: { requestsPerSecond: 10, burstSize: 20 },
          llm: { requestsPerSecond: 3, burstSize: 6 },
        },
      };
      mockExistsSyncReturnValue = true;
      mockReadFileSyncReturnValue = JSON.stringify(validConfig);

      const config = loadConfig();
      expect(config).toMatchObject(validConfig);
      expect(config?.ratelimit?.embedding?.requestsPerSecond).toBe(10);
      expect(config?.ratelimit?.embedding?.burstSize).toBe(20);
      expect(config?.ratelimit?.llm?.requestsPerSecond).toBe(3);
      expect(config?.ratelimit?.llm?.burstSize).toBe(6);
    });

    it('should return LLMConfig with partial ratelimit settings', () => {
      const validConfig = {
        provider: 'gemini' as const,
        apiKey: 'test-api-key',
        ratelimit: {
          embedding: { requestsPerSecond: 10 },
        },
      };
      mockExistsSyncReturnValue = true;
      mockReadFileSyncReturnValue = JSON.stringify(validConfig);

      const config = loadConfig();
      expect(config).toMatchObject(validConfig);
      expect(config?.ratelimit?.embedding?.requestsPerSecond).toBe(10);
      expect(config?.ratelimit?.llm).toBeUndefined();
    });
  });

  describe('when config uses providers[] without a top-level apiKey', () => {
    it('should load successfully', () => {
      const validConfig = {
        provider: 'gemini' as const,
        providers: [
          { apiKey: 'key1', model: 'gemma-4-31b-it' },
          { apiKey: 'key2', model: 'gemma-4-26b-a4b-it' },
        ],
      };
      mockExistsSyncReturnValue = true;
      mockReadFileSyncReturnValue = JSON.stringify(validConfig);

      const config = loadConfig();
      expect(config).toMatchObject(validConfig);
    });
  });

  describe('when config file exists but is invalid', () => {
    it('should return null and log warning for invalid JSON', () => {
      mockExistsSyncReturnValue = true;
      mockReadFileSyncReturnValue = 'invalid json{';

      const warnSpy = mock(() => {});
      const originalWarn = console.warn;
      console.warn = warnSpy;

      try {
        const config = loadConfig();
        expect(config).toBeNull();
      } finally {
        console.warn = originalWarn;
      }
    });

    it('should return null when provider is missing', () => {
      const invalidConfig = {
        apiKey: 'test-api-key',
        model: 'gemini-2.0-flash',
      };
      mockExistsSyncReturnValue = true;
      mockReadFileSyncReturnValue = JSON.stringify(invalidConfig);

      const config = loadConfig();
      expect(config).toBeNull();
    });

    it('should return null when apiKey is missing', () => {
      const invalidConfig = {
        provider: 'gemini',
        model: 'gemini-2.0-flash',
      };
      mockExistsSyncReturnValue = true;
      mockReadFileSyncReturnValue = JSON.stringify(invalidConfig);

      const config = loadConfig();
      expect(config).toBeNull();
    });

    it('should return null for empty provider string', () => {
      const invalidConfig = {
        provider: '',
        apiKey: 'test-api-key',
      };
      mockExistsSyncReturnValue = true;
      mockReadFileSyncReturnValue = JSON.stringify(invalidConfig);

      const config = loadConfig();
      expect(config).toBeNull();
    });

    it('should return null for unknown provider', () => {
      const invalidConfig = {
        provider: 'unknown',
        apiKey: 'test-api-key',
      };
      mockExistsSyncReturnValue = true;
      mockReadFileSyncReturnValue = JSON.stringify(invalidConfig);

      const config = loadConfig();
      expect(config).toBeNull();
    });
  });
});

describe('createProvider', () => {
  describe('valid configurations', () => {
    it('should create GeminiProvider with provider, apiKey and default model', async () => {
      const config: LLMConfig = {
        provider: 'gemini',
        apiKey: 'test-api-key',
      };

      const provider = await createProvider(config);

      expect(provider).toBeDefined();
      expect(provider.complete).toBeDefined();
      expect(typeof provider.complete).toBe('function');
    });

    it('should create GeminiProvider with provider, apiKey and specified model', async () => {
      const config: LLMConfig = {
        provider: 'gemini',
        apiKey: 'test-api-key',
        model: 'gemini-2.0-flash',
      };

      const provider = await createProvider(config);

      expect(provider).toBeDefined();
      expect(provider.complete).toBeDefined();
      expect(typeof provider.complete).toBe('function');
    });

    it('should create ZAIProvider with zai provider', async () => {
      const config: LLMConfig = {
        provider: 'zai',
        apiKey: 'test-api-key',
      };

      const provider = await createProvider(config);

      expect(provider).toBeDefined();
      expect(provider.complete).toBeDefined();
      expect(typeof provider.complete).toBe('function');
    });

    it('should create provider with correct structure for use', async () => {
      const config: LLMConfig = {
        provider: 'gemini',
        apiKey: 'test-key',
      };

      const provider = await createProvider(config);

      expect(provider).toBeDefined();
      expect(typeof provider.complete).toBe('function');
      expect(provider.complete.length).toBeGreaterThanOrEqual(0);
    });
  });

  describe('providers array (round-robin)', () => {
    it('should create a rotating provider when providers[] is set', async () => {
      const config: LLMConfig = {
        provider: 'gemini',
        providers: [
          { apiKey: 'key1', model: 'gemma-4-31b-it' },
          { apiKey: 'key1', model: 'gemma-4-26b-a4b-it' },
          { apiKey: 'key2', model: 'gemma-4-31b-it' },
        ],
      };

      const provider = await createProvider(config);

      expect(provider).toBeDefined();
      expect(typeof provider.complete).toBe('function');
      // RoundRobinProvider exposes the same LLMProvider surface
      expect(provider.constructor.name).toBe('RoundRobinProvider');
    });

    it('should accept an optional email on each provider entry', async () => {
      const config: LLMConfig = {
        provider: 'gemini',
        providers: [
          { email: 'baleentest001@gmail.com', apiKey: 'key1', model: 'gemma-4-31b-it' },
          { email: 'baleentest001@gmail.com', apiKey: 'key2', model: 'gemma-4-26b-a4b-it' },
        ],
      };

      const provider = await createProvider(config);

      expect(provider.constructor.name).toBe('RoundRobinProvider');
      expect(config.providers?.[0].email).toBe('baleentest001@gmail.com');
    });

    it('should throw when providers[] is empty', async () => {
      const config: LLMConfig = {
        provider: 'gemini',
        providers: [],
      };

      await expect(createProvider(config)).rejects.toThrow();
    });
  });

  describe('error handling', () => {
    it('should throw error when apiKey is missing', async () => {
      const config = {
        provider: 'gemini',
        model: 'gemini-2.0-flash',
      } as LLMConfig;

      await expect(createProvider(config)).rejects.toThrow('requires an apiKey');
    });

    it('should throw error when apiKey is empty string', async () => {
      const config: LLMConfig = {
        provider: 'gemini',
        apiKey: '',
      };

      await expect(createProvider(config)).rejects.toThrow('requires an apiKey');
    });
  });
});

describe('embedding config', () => {
  afterEach(() => {
    __setConfigFileDepsForTests(null);
  });

  it('reads embedding.maxConcurrency from the config file', () => {
    __setConfigFileDepsForTests({
      existsSync: () => true,
      readFileSync: (() => JSON.stringify({
        provider: 'gemini',
        apiKey: 'k',
        embedding: { maxConcurrency: 7 },
      })) as never,
    });
    expect(loadConfig()?.embedding?.maxConcurrency).toBe(7);
  });

  it('leaves embedding undefined when the config omits it', () => {
    __setConfigFileDepsForTests({
      existsSync: () => true,
      readFileSync: (() => JSON.stringify({ provider: 'gemini', apiKey: 'k' })) as never,
    });
    expect(loadConfig()?.embedding).toBeUndefined();
  });

  it('DEFAULT_EMBEDDING_MAX_CONCURRENCY is 4', () => {
    expect(DEFAULT_EMBEDDING_MAX_CONCURRENCY).toBe(4);
  });
});

File: internal/llm/config_test.go
package llm

import (
	"strings"
	"testing"
)

// setStubConfig wires SetConfigFileDepsForTests so LoadConfig sees the given
// content (exists=true) or a missing file (exists=false).
func setStubConfig(t *testing.T, exists bool, content string) {
	t.Helper()
	SetConfigFileDepsForTests(
		func(string) bool { return exists },
		func(string) ([]byte, error) { return []byte(content), nil },
	)
	t.Cleanup(func() { SetConfigFileDepsForTests(nil, nil) })
}

func TestLoadConfig_MissingFileReturnsNil(t *testing.T) {
	setStubConfig(t, false, "")
	cfg, err := LoadConfig()
	if err != nil {
		t.Fatalf("err = %v", err)
	}
	if cfg != nil {
		t.Fatalf("cfg = %+v, want nil", cfg)
	}
}

func TestLoadConfig_ValidGemini(t *testing.T) {
	setStubConfig(t, true, `{"provider":"gemini","apiKey":"test-api-key","model":"gemini-2.0-flash"}`)
	cfg, err := LoadConfig()
	if err != nil || cfg == nil {
		t.Fatalf("cfg=%v err=%v", cfg, err)
	}
	if cfg.Provider != ProviderGemini || cfg.APIKey != "test-api-key" || cfg.Model != "gemini-2.0-flash" {
		t.Fatalf("cfg = %+v", cfg)
	}
}

func TestLoadConfig_ValidZAI(t *testing.T) {
	setStubConfig(t, true, `{"provider":"zai","apiKey":"test-api-key","model":"glm-4.7"}`)
	cfg, err := LoadConfig()
	if err != nil || cfg == nil {
		t.Fatalf("cfg=%v err=%v", cfg, err)
	}
	if cfg.Provider != ProviderZAI {
		t.Fatalf("provider = %q", cfg.Provider)
	}
}

func TestLoadConfig_RateLimitParsed(t *testing.T) {
	setStubConfig(t, true, `{"provider":"gemini","apiKey":"k","ratelimit":{"embedding":{"requestsPerSecond":10,"burstSize":20},"llm":{"requestsPerSecond":3,"burstSize":6}}}`)
	cfg, err := LoadConfig()
	if err != nil || cfg == nil {
		t.Fatalf("cfg=%v err=%v", cfg, err)
	}
	if cfg.RateLimit == nil || cfg.RateLimit.Embedding == nil || cfg.RateLimit.LLM == nil {
		t.Fatalf("ratelimit = %+v", cfg.RateLimit)
	}
	if *cfg.RateLimit.Embedding.RequestsPerSecond != 10 || *cfg.RateLimit.Embedding.BurstSize != 20 {
		t.Fatalf("embedding = %+v", cfg.RateLimit.Embedding)
	}
	if *cfg.RateLimit.LLM.RequestsPerSecond != 3 || *cfg.RateLimit.LLM.BurstSize != 6 {
		t.Fatalf("llm = %+v", cfg.RateLimit.LLM)
	}
}

func TestLoadConfig_PartialRateLimit(t *testing.T) {
	setStubConfig(t, true, `{"provider":"gemini","apiKey":"k","ratelimit":{"embedding":{"requestsPerSecond":10}}}`)
	cfg, err := LoadConfig()
	if err != nil || cfg == nil {
		t.Fatalf("cfg=%v err=%v", cfg, err)
	}
	if *cfg.RateLimit.Embedding.RequestsPerSecond != 10 {
		t.Fatalf("rps = %v", cfg.RateLimit.Embedding.RequestsPerSecond)
	}
	if cfg.RateLimit.LLM != nil {
		t.Fatalf("llm should be nil, got %+v", cfg.RateLimit.LLM)
	}
}

func TestLoadConfig_ProvidersWithoutTopLevelAPIKey(t *testing.T) {
	setStubConfig(t, true, `{"provider":"gemini","providers":[{"apiKey":"key1","model":"gemma-4-31b-it"},{"apiKey":"key2"}]}`)
	cfg, err := LoadConfig()
	if err != nil || cfg == nil {
		t.Fatalf("cfg=%v err=%v", cfg, err)
	}
	if len(cfg.Providers) != 2 || cfg.Providers[0].APIKey != "key1" {
		t.Fatalf("providers = %+v", cfg.Providers)
	}
}

func TestLoadConfig_InvalidCases(t *testing.T) {
	cases := []struct {
		name    string
		content string
	}{
		{"invalid json", "invalid json{"},
		{"missing provider", `{"apiKey":"k","model":"gemini-2.0-flash"}`},
		{"missing apiKey", `{"provider":"gemini","model":"gemini-2.0-flash"}`},
		{"empty provider string", `{"provider":"","apiKey":"k"}`},
		{"unknown provider", `{"provider":"unknown","apiKey":"k"}`},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			setStubConfig(t, true, tc.content)
			cfg, err := LoadConfig()
			if err != nil {
				t.Fatalf("err = %v", err)
			}
			if cfg != nil {
				t.Fatalf("cfg = %+v, want nil", cfg)
			}
		})
	}
}

func TestCreateProvider_GeminiDefaultModel(t *testing.T) {
	p, err := CreateProvider(Config{Provider: ProviderGemini, APIKey: "k"})
	if err != nil {
		t.Fatalf("CreateProvider: %v", err)
	}
	g, ok := p.(*GeminiProvider)
	if !ok {
		t.Fatalf("type = %T", p)
	}
	if g.model != "gemini-2.0-flash" {
		t.Fatalf("model = %q", g.model)
	}
}

func TestCreateProvider_GeminiSpecifiedModel(t *testing.T) {
	p, err := CreateProvider(Config{Provider: ProviderGemini, APIKey: "k", Model: "gemma-4-31b-it"})
	if err != nil {
		t.Fatalf("CreateProvider: %v", err)
	}
	if p.(*GeminiProvider).model != "gemma-4-31b-it" {
		t.Fatalf("model = %q", p.(*GeminiProvider).model)
	}
}

func TestCreateProvider_ZAI(t *testing.T) {
	p, err := CreateProvider(Config{Provider: ProviderZAI, APIKey: "k"})
	if err != nil {
		t.Fatalf("CreateProvider: %v", err)
	}
	z, ok := p.(*ZAIProvider)
	if !ok {
		t.Fatalf("type = %T", p)
	}
	if z.model != "glm-4.5-air" {
		t.Fatalf("model = %q", z.model)
	}
}

func TestCreateProvider_RoundRobin(t *testing.T) {
	p, err := CreateProvider(Config{
		Provider: ProviderGemini,
		Providers: []ProviderEntry{
			{APIKey: "key1", Model: "gemma-4-31b-it"},
			{APIKey: "key1", Model: "gemma-4-26b-a4b-it"},
			{APIKey: "key2", Model: "gemma-4-31b-it"},
		},
	})
	if err != nil {
		t.Fatalf("CreateProvider: %v", err)
	}
	rr, ok := p.(*RoundRobinProvider)
	if !ok {
		t.Fatalf("type = %T, want *RoundRobinProvider", p)
	}
	if len(rr.providers) != 3 {
		t.Fatalf("providers = %d", len(rr.providers))
	}
}

func TestCreateProvider_EmptyProvidersErrors(t *testing.T) {
	_, err := CreateProvider(Config{Provider: ProviderGemini, Providers: []ProviderEntry{}})
	if err == nil {
		t.Fatal("expected error for empty providers[]")
	}
}

func TestCreateProvider_MissingAPIKeyErrors(t *testing.T) {
	_, err := CreateProvider(Config{Provider: ProviderGemini})
	if err == nil || !strings.Contains(err.Error(), "requires an apiKey") {
		t.Fatalf("err = %v, want 'requires an apiKey'", err)
	}
}

func TestCreateProvider_UnknownProviderErrors(t *testing.T) {
	_, err := CreateProvider(Config{Provider: "bogus", APIKey: "k"})
	if err == nil || !strings.Contains(err.Error(), "Unknown provider") {
		t.Fatalf("err = %v, want 'Unknown provider'", err)
	}
}
